providers/common/ciphers/aes_gcm.c

```diff
@@ -0,0 +1,565 @@
+}
+
+static int aes_gcm_ctx_get_params(void *vctx, OSSL_PARAM params[])
+{
+    PROV_AES_GCM_CTX *ctx = (PROV_AES_GCM_CTX *)vctx;
+    OSSL_PARAM *p;
+    size_t sz;
+
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_KEYLEN);
+    if (p != NULL) {
+        if (!OSSL_PARAM_set_int(p, ctx->keylen)) {
+            PROVerr(PROV_F_AES_GCM_CTX_GET_PARAMS, PROV_R_FAILED_TO_SET_PARAMETER);
+            return 0;
+        }
+    }
+
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_IV);
+    if (p != NULL) {
+        if (ctx->iv_gen != 1 && ctx->iv_gen_rand != 1)
+            return 0;
+        if (ctx->ivlen != (int)p->data_size) {
+            PROVerr(PROV_F_AES_GCM_CTX_GET_PARAMS, PROV_R_INVALID_IVLEN);
+            return 0;
+        }
+        if (!OSSL_PARAM_set_octet_string(p, ctx->iv, ctx->ivlen)) {
+            PROVerr(PROV_F_AES_GCM_CTX_GET_PARAMS, PROV_R_FAILED_TO_SET_PARAMETER);
+            return 0;
+        }
+    }
+
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_AEAD_TLS1_AAD_PAD);
+    if (p != NULL) {
+        if (!OSSL_PARAM_set_size_t(p, ctx->tls_aad_pad_sz)) {
+            PROVerr(PROV_F_AES_GCM_CTX_GET_PARAMS, PROV_R_FAILED_TO_SET_PARAMETER);
+            return 0;
+        }
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_AEAD_TAG);
+    if (p != NULL) {
+        sz = p->data_size;
+        if (sz == 0 || sz > EVP_GCM_TLS_TAG_LEN || !ctx->enc || ctx->taglen < 0) {
+            PROVerr(PROV_F_AES_GCM_CTX_GET_PARAMS, PROV_R_INVALID_TAG);
+            return 0;
+        }
+        if (!OSSL_PARAM_set_octet_string(p, ctx->buf, sz)) {
+            PROVerr(PROV_F_AES_GCM_CTX_GET_PARAMS, PROV_R_FAILED_TO_SET_PARAMETER);
+            return 0;
+        }
+    }
+    return 1;
+}
+
+static int aes_gcm_ctx_set_params(void *vctx, const OSSL_PARAM params[])
+{
+    PROV_AES_GCM_CTX *ctx = (PROV_AES_GCM_CTX *)vctx;
+    const OSSL_PARAM *p;
+    size_t sz;
+    void *vp;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_AEAD_TAG);
+    if (p != NULL) {
+        vp = ctx->buf;
+        if (!OSSL_PARAM_get_octet_string(p, &vp, EVP_GCM_TLS_TAG_LEN, &sz)) {
+            PROVerr(PROV_F_AES_GCM_CTX_SET_PARAMS, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+        if (sz == 0 || ctx->enc) {
+            PROVerr(PROV_F_AES_GCM_CTX_SET_PARAMS, PROV_R_INVALID_TAG);
+            return 0;
+        }
+        ctx->taglen = sz;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_AEAD_IVLEN);
+    if (p != NULL) {
+        if (!OSSL_PARAM_get_size_t(p, &sz)) {
+            PROVerr(PROV_F_AES_GCM_CTX_SET_PARAMS, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+        if (sz == 0 || sz > sizeof(ctx->iv)) {
+            PROVerr(PROV_F_AES_GCM_CTX_SET_PARAMS, PROV_R_INVALID_IVLEN);
+            return 0;
+        }
+        ctx->ivlen = sz;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_AEAD_TLS1_AAD);
```

Comment:
I wonder whether we should actually implement TLS-GCM as an entirely different cipher rather than as a special case version of standard GCM?

Reply:
That might be a long term goal, maybe not for this PR?
I might be difficult to do at the moment since the deprecated code still needs to work for a while. I am also not sure if there are any cases where it might swap between the 2 bits of code inside tls.

Reply:
> TLS-GCM as an entirely different cipher

only if still possible to handle whole packet in one shot and not enforce an init-update(wait for more)-final sequence.

Reply:
Yeah...out of scope of this PR - but it seems to make sense to me.

> only if its still possible to handle whole packet in one shot and not enforce an init-update(wait for more)-final sequence.

Sure